### ISSUE TITLE
Regression: dynamic coercion of Model will clobber property values on update

### ISSUE THREAD

**Issue by fabien** [strongloop/loopback#1806]
I finally went through a full update cycle with my apps, and in updating from 2.18.0 to 2.23.0 I discovered the following undesired (and BC breaking) behavior:

With a property that has some default value (either static or a fn/dynamic value):
```js
"properties": {
  "foo": {
    "type": "string",
    "default": "bar"
  }
}
```

```js
// create receives a new instance, with defaults filled in
// which we'll overwrite by our new data
this.request.post('/api/things')
  .send({ foo: 'other' }).end(function(err, res) {
  res.body.foo.should.equal('other');
  // id: 1234567890
  next();
});

// updateAttributes now receives a new instance, with defaults filled in
// but this time, we don't have any value for 'foo' set explicitly
// and thus the default will be used for the update(!)
this.request.put('/api/things/1234567890')
  .send({ biz: 'baz' }).end(function(err, res) {
  res.body.biz.should.equal('baz');
  res.body.foo.should.equal('other'); // Fail: overwritten by default: 'bar'
  next();
});
```
I think it's related to this commit: https://github.com/strongloop/strong-remoting/commit/34b0665841d763944dca3e9b4b2033366ea8a0c0

/cc @bajtos @ritch 

---

See https://github.com/strongloop/loopback/issues/1806#issuecomment-217393388 for the proposed solution.

 - [ ]  Implement support for `arg.model` in loopback-swagger
 - [ ] Rework remoting metadata in LoopBack, use `{ type: 'object', model: modelName }` everywhere. Make sure to add unit-tests reproducing the problem and then demonstrating that the problem is fixed.

The last item from the proposal is tracked by a standalone issue: https://github.com/strongloop/loopback/issues/2313

> Fix swagger metadata for updateAttributes and friends to correctly indicate all properties are optional. Check other schema parts like "default" values, we may need to remove them too.

**Comment by fabien** [comment]
I was able to identify where the issue exactly happens, but the fix is unsatisfactory because it prevents dynamic (object) type conversion. 

However, it's extremely difficult to not end up with the issue here - how can we declare that the coerced values are in fact just *defaults* and that they should *not* overwrite any existing data on update?

**Comment by fabien** [comment]
@raymondfeng any thoughts on this? The fix I mention over at https://github.com/strongloop/strong-remoting/issues/254 works fine, but I'm not sure if this has any actual drawbacks at the moment.

**Comment by raymondfeng** [comment]
@fabien Should we introduce an option such as `applyDefaults` (default to true) for the model constructor so that the coercion can skip the defaults?

**Comment by fabien** [comment]
@raymondfeng sounds reasonable, yes - then again, I don't think strong-remoting is even aware of Loopback models, let alone the passing of such options to a constructor?

**Comment by fabien** [comment]
@bajtos +1 for label: major

**Comment by raymondfeng** [comment]
@fabien It starts to make me wonder if we should introduce a static method on the model class to handle the coercion, such as `MyModel.coerce(rawValue)`.

We have a few different flavors in creating a model instance.

- Construct a model instance, applying defaults and validating properties
- Loading raw json object from DB or REST into a model instance as-is (No defaults will be applied, with or without validation)

**Comment by fabien** [comment]
Having a public method like that would certainly be great (I found myself wishing for this before), and it looks like this will resolve the issue.

**Comment by fabien** [comment]
@raymondfeng digging further, I did find `applyDefaultValues` to be already implemented here: https://github.com/strongloop/loopback-datasource-juggler/blob/934786200b0180ed3d3cf45f78986b2019511e79/lib/model.js#L79

Which suggests that we can change the following: https://github.com/strongloop/loopback/blob/master/lib/model.js#L131

Into:

```
    RemoteObjects.convert(typeName, function(val) {
      return val ? new ModelCtor(val, { applyDefaultValues: false }) : val;
    });
```

Unfortunately I wasn't able to test this, because of multiple test failures (out of the box) with the latest pull of `loopback`. This is a blocking issue for me - your feedback is welcome! Thanks.

**Comment by fabien** [comment]
Update: the above does not work at all. I'm really starting to wonder if we actually want this coercion to be applied when the httpContext is being created. We already have a proper `sharedCtor` to handle this for us. Think about it for a moment:

The httpContext coerced model instance will always be a 'new' model; I can't really think of any use for such a model to be made available in coerced form at this point of the remoting invocation. We feed plain objects to our objects anyway (and whenever we don't, we convert them internally using `.toObject` first (see `ModelBaseClass.prototype._initProperties`).

The issue here has only been introduced recently, when strong-remoting was updated to do this coercion. Can we please revert to the former behavior? Using the workaround mentioned?

**Comment by fabien** [comment]
@raymondfeng please see referenced PR's. Thanks.

**Comment by fabien** [comment]
@raymondfeng ping

**Comment by bajtos** [comment]
Hey @fabien, thank you for the detailed analysis. In principle, your conclusion that strong-remoting should not coerce model data to model instances makes sense to me and I fully agree. Before implementing that change, I'd like to hear from @raymondfeng and @ritch why that recent change was made - perhaps we (you and me) are both missing something important here?

**Comment by bajtos** [comment]
Actually, as I am thinking about this some more, perhaps we should not label accept arguments with `type:modelName` instead? For example, for `updateAttributes`, the argument is not a valid model, it's a subset of model properties. I'm throwing this in just as an idea to consider.

**Comment by bajtos** [comment]
BTW the type converter was registration was added as part of LB 2.0 work, see https://github.com/fabien/loopback/commit/918596f0353b726d4b0da7142a6b8e0aefa94b11

**Comment by fabien** [comment]
@bajtos thanks for the update. It's strange that this has been committed as part of LB 2.0 such a long time ago. I did not expect that, because this issue has only surfaced recently for me. All this time the given test cases passed, all the way until my recent update from 2.18.0 to 2.23.0.

There must be something else involved, and I suspect this to be in `strong-demoting` instead.

**Comment by bajtos** [comment]
> There must be something else involved, and I suspect this to be in strong-remoting instead.

That's entirely possible. Could you please share the failing test case with us so that we can isolate the particular commit which broke it? Or perhaps could you do that search yourself?

**Comment by fabien** [comment]
@bajtos as initially mentioned I think it's related to https://github.com/strongloop/strong-remoting/commit/34b0665841d763944dca3e9b4b2033366ea8a0c0

A sample failing test case is also outlined at the top of this issue ;)

**Comment by fabien** [comment]
@raymondfeng @ritch @bajtos - any updates?

**Comment by bajtos** [comment]
I personally think this needs a bit of research first: review recent changes (e.g. strongloop/strong-remoting@34b0665), proposed changes (e.g. strongloop/strong-remoting#264) and related GH issues that are still opened; build a better understanding of various edge cases that we need to support. Then write down a proposal how to fix this properly, in way that would hopefully prevent more bugs like this one popping up in the future. (I know we cannot prevent all bugs, but we should at least make the best effort.)

**Comment by shaharz** [comment]
Any updates on this? This is a showstopper bug for me and unfortunately it looks like this one requires an intimate knowledge of Loopback internals which I currently lack, otherwise I would gladly fix it myself.

I put together a test case before I found out about this issue, hopefully it can help you come up with a fix or at least a workaround: https://github.com/shaharz/loopback-sandbox/tree/related-default-bug

Thanks!

**Comment by bajtos** [comment]
Hello, sorry for the long delay. I have started to look into ways how to fix this issue "properly", now I would like to share my thoughts and get your feedback before I start making any changes.

I am afraid I am not able to reproduce the issue when calling `PUT /api/mymodel/{id}` directly. This is actually not very surprising, as the remoting metadata for `updateAttributes` specifies that the type of the input argument `data` is a plain object, see [lib/persisted-model.js](https://github.com/strongloop/loopback/blob/1559db2ca3f5cf75606a09a806428140e9c967d6/lib/persisted-model.js#L697).

I am able to reproduce the problem when calling `updateById` via `hasMany` relation. This is again not very surprising, because the remoting metadata specifies that the input argument `data` should be a model instance, see [lib/model.js](https://github.com/strongloop/loopback/blob/1559db2ca3f5cf75606a09a806428140e9c967d6/lib/model.js#L547).

To fix this issue and remove the regression:
**I am proposing to modify remoting metadata for relations to correctly specify that the input `data` argument should be a plain object.**

However, such change would have negative consequences on the swagger specification produced by loopback-swagger, as it will either no longer provide metadata about the request body, or it will think the request body should have the type of the model the modified instance belongs to. (See this hack in loopback-swagger: [lib/specgen/route-helper.js](https://github.com/strongloop/loopback-swagger/blob/4b2cf3a7e642331b55cb85bbd34ca5e60292a939/lib/specgen/route-helper.js#L280-L282)).

To prevent that, I am proposing the following changes:
 - Add a new remoting metadata property called `model` that will be used to inform swagger specgen about the type. `{ arg: 'data', type: 'object', model: 'MyRelatedModel' }`. This new property will be ignored by strong-remoting, it will continue using `type:'object'` as before.
 - In LoopBack, modify all remoting metadata for default CRUD/DAO methods to correctly use this new feature.

Additionally, it would be nice to fix the following issue that is loosely related: ATM, the `update` methods perform a partial update (patch), but the description of the request does not convey this information. Specifically, if a model has a required property, this property may be omitted from an update (patch) request, but the swagger spec says it's still required. I think the solution is to register a new type (model, definition) that will have all properties of the original model, but none of them will be marked as required. In the future, we can extend this new type to describe extended operators like `$inc` too, but that's out of scope of the initial implementation.

Summary of proposed steps:
 1. Implement support for `arg.model` in loopback-swagger
 2. Rework remoting metadata in LoopBack, use `{ type: 'object', model: modelName }` everywhere
 3. Fix swagger metadata for `updateAttributes` and friends to correctly indicate all properties are optional. Check other schema parts like "default" values, we may need to remove them too.

@fabien @raymondfeng @shaharz thoughts?

**Comment by bajtos** [comment]
To prevent confusion, when I say the current behaviour is not surprising, I mean the behaviour matches what is expressed in the source code. I fully agree that the current behaviour is surprising for LoopBack users, it's a major bug and it should be fixed.

**Comment by bajtos** [comment]
> I am proposing to modify remoting metadata for relations to correctly specify that the input data argument should be a plain object.

Let me describe a bit more why I would like to make this change:

 - We already do the same for non-relation CRUD methods.
 - All juggler methods support plain object for `data` argument.
 - Creating a model instance before calling juggler methods adds unnecessary performance overhead.

**Comment by shaharz** [comment]
@bajtos sounds good to me.

**Comment by ritch** [comment]
```js
{ arg: 'data', type: 'object', model: 'MyRelatedModel' }
```

I think this gets us close... but what the real issue is that we aren't properly specifying the expected schema for the `data` object in `updateById`. This is really a limitation of strong-remoting's "type". It expects you to specify a model or JSON type. This is different than what swagger expects. Which is json-schema.

I'd really like to add something like `schema` to strong-remoting or loopback that would allow you to specify json schema as well as the argument's type. This is a good sign that we should bump the priority of https://github.com/strongloop/loopback/issues/2109

I'm 👍 for this as an incremental change, as I think this is future proof if we add a `schema` property to the argument definition or support json schema in loopback.

**Comment by bajtos** [comment]
> Fix swagger metadata for updateAttributes and friends to correctly indicate all properties are optional. Check other schema parts like "default" values, we may need to remove them too.

Moved to https://github.com/strongloop/loopback/issues/2726

**Comment by bajtos** [comment]
Released in `loopback@2.33.0`.
